### PR TITLE
Fixed issue #19173: encrypted file upload question - delete file results in crash

### PR DESCRIPTION
**Pull request by mfavetti**
Fixed issue #19173: encrypted file upload question - delete file results in crash

[https://bugs.limesurvey.org/view.php?id=19173](https://bugs.limesurvey.org/view.php?id=19173)

### CODE REVIEW

File: application/models/Response.php [hunk 1]
```diff
@@ -109,14 +109,20 @@ public function getFilesAndSqga($sQID = 0)
             ->findAllByAttributes($aConditions);
         $files = array();
         foreach ($aQuestions as $question) {
+
+            $aQuestion = array(
```
Comment (Shnoulle) [inline application/models/Response.php:113]: Maybe more clear with `$encrypted = $question->encrypted === 'Y',`

File: application/models/Response.php [hunk 2]
```diff
@@ -109,14 +109,20 @@ public function getFilesAndSqga($sQID = 0)
...
             $field = $question->sid . 'X' . $question->gid . 'X' . $question->qid;
             $fieldDataJson = $this->getAttribute($field);
-            if ($question->encrypted === 'Y') {
+            if ($aQuestion['encrypted']) {
                 $fieldDataJson = self::decryptSingle($fieldDataJson);
             }
             $fieldData = json_decode(stripslashes((string) $fieldDataJson), true);
             if (is_array($fieldData)) {
-                $files[$field] = $fieldData;
+                $aQuestion['files'] = $fieldData;
```
Comment (Shnoulle) [inline application/models/Response.php:124]: And here : 
```

$files[$field] = [
    'file'= $fieldData,
    'encrypted' => $encrypted
];
```

**Review by Shnoulle — COMMENTED** [review]
Code is OK,

but whole system seems really complex for objectives 

```
./application$ grep -r "getFilesAndSqga" *
models/Response.php:    public function getFilesAndSqga($sQID = 0)
models/Response.php:        $filesData = $this->getFilesAndSqga();
./application$ grep -r "deleteFilesAndFilename" *
controllers/ResponsesController.php:                [$success, $errors] = $response->deleteFilesAndFilename();
models/Response.php:    public function deleteFilesAndFilename()

```

**Comment by mfavetti** [comment]
> Code is OK,
> 
> but whole system seems really complex for objectives
> 
> ```
> ./application$ grep -r "getFilesAndSqga" *
> models/Response.php:    public function getFilesAndSqga($sQID = 0)
> models/Response.php:        $filesData = $this->getFilesAndSqga();
> ./application$ grep -r "deleteFilesAndFilename" *
> controllers/ResponsesController.php:                [$success, $errors] = $response->deleteFilesAndFilename();
> models/Response.php:    public function deleteFilesAndFilename()
> ```

Can't really speak to existing implementation...

Do you mean that each function is only used in one place, so no need to abstract them? Like it's better to just use the function's code directly where called?

**Comment by Shnoulle** [comment]
> > Code is OK,
> > but whole system seems really complex for objectives
> > ```
> > ./application$ grep -r "getFilesAndSqga" *
> > models/Response.php:    public function getFilesAndSqga($sQID = 0)
> > models/Response.php:        $filesData = $this->getFilesAndSqga();
> > ./application$ grep -r "deleteFilesAndFilename" *
> > controllers/ResponsesController.php:                [$success, $errors] = $response->deleteFilesAndFilename();
> > models/Response.php:    public function deleteFilesAndFilename()
> > ```
> 
> Can't really speak to existing implementation...
> 
> Do you mean that each function is only used in one place, so no need to abstract them? Like it's better to just use the function's code directly where called?

Really unsure, but when reviewing : i start to make some comment for code optimization but after reviewing more : i start to understand why you di it like this.

Here : i really don't understand the point to use `getFilesAndSqga`

Just doing a 

```
foreach ($aQuestions as $question) {
    $field = $question->sid . 'X' . $question->gid . 'X' . $question->qid;
    $fieldDataJson = $this->getAttribute($field);
    if ($question->encrypted === 'Y') {
        $fieldDataJson = self::decryptSingle($fieldDataJson);
    }
    $fieldData = json_decode(urldecode((string) $fieldDataJson), true);
    if (is_array($fieldData)) {
        // Do the delete action
    }
}
```


Seems more clear and clean.

**Comment by Shnoulle** [comment]
PS : remind to activate phpcs on your code.

**Comment by kevin-foster-uk** [comment]
Fixed in: https://github.com/LimeSurvey/LimeSurvey/pull/3564

**Comment by mfavetti** [comment]
> PS : remind to activate phpcs on your code.

i will look into this, i didn't know about it

**Comment by Shnoulle** [comment]
> > PS : remind to activate phpcs on your code.
> 
> i will look into this, i didn't know about it

What is your editor ?

**Comment by mfavetti** [comment]
> > > PS : remind to activate phpcs on your code.
> > 
> > 
> > i will look into this, i didn't know about it
> 
> What is your editor ?

sublime

**Comment by Shnoulle** [comment]
> > > > PS : remind to activate phpcs on your code.
> > > 
> > > 
> > > i will look into this, i didn't know about it
> > 
> > 
> > What is your editor ?
> 
> sublime

Seems there are plugin : https://github.com/benmatselby/sublime-phpcs

Currently i use geany for simple plugin, but use vscodium automatic fix before push (when i remind …)

**Pull request closed without merging**